### PR TITLE
Support HAProxy proxy protocol for broker and proxy

### PR DESCRIPTION
**Pull request by codelipenghui**
### Motivation

Currently, if enable the proxy in the pulsar cluster and client connect to the cluster through the proxy, when we get topic stats, the consumer address and producer address are not the real client address. This PR fix this problem by leverage HAProxy proxy protocol since this is a more general approach.  more details about the proxy protocol see 
https://www.haproxy.com/blog/haproxy/proxy-protocol/
https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/enable-proxy-protocol.html
http://www.haproxy.org/download/1.8/doc/proxy-protocol.txt

### Modifications

Allow enable proxy protocol on proxy or broker.

### Verifying this change

Tests added

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (yes)

### CODE REVIEW

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java [hunk 1]
```diff
@@ -169,6 +171,11 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
 
     @Override
     public void channelRead(final ChannelHandlerContext ctx, Object msg) throws Exception {
+        System.out.println(haProxyMessage);
```
Comment (sijie) [inline pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java:174]: remove this?

**Comment by eolivelli** [comment]
@diegosalvi you have recently implemented this feature on EmailSuccess, can you please take a look?

File: conf/broker.conf [hunk 2]
```diff
@@ -55,6 +55,9 @@ advertisedAddress=
 # The Default value is absent, the broker uses the first listener as the internal listener.
 # internalListenerName=
 
+# Enable or disable the proxy protocol.
+proxyProtocolEnabled=false
```
Comment (merlimat) [inline conf/broker.conf:59]: This naming can be confusing since we already have a "proxy" for Pulsar protocol

Comment (codelipenghui) [reply conf/broker.conf:59]: How about changing it to `haProxyProtocolEnabled`?

Comment (sijie) [reply conf/broker.conf:59]: `haProxyProtocol` looks much better than `proxyProtocol`.

Comment (codelipenghui) [reply conf/broker.conf:59]: Ok, sounds good.

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java [hunk 3]
```diff
@@ -143,10 +150,56 @@ protected void initChannel(SocketChannel ch) throws Exception {
...
+                } else {
+                    if (inboundChannel.remoteAddress() instanceof InetSocketAddress) {
+                        InetSocketAddress clientAddress = (InetSocketAddress) inboundChannel.remoteAddress();
+                        String sourceAddress = clientAddress.getAddress().getHostAddress();
+                        int sourcePort = clientAddress.getPort();
+                        if (outboundChannel.localAddress() instanceof InetSocketAddress) {
+                            InetSocketAddress proxyAddress = (InetSocketAddress) inboundChannel.remoteAddress();
+                            String destinationAddress = proxyAddress.getAddress().getHostAddress();
+                            int destinationPort = proxyAddress.getPort();
+                            HAProxyMessage msg = new HAProxyMessage(HAProxyProtocolVersion.V1, HAProxyCommand.PROXY,
```
Comment (diegosalvi) [inline pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java:165]: netty-haproxy can decode binary v2 version too (but doesn't implements encoding).
We could add another encoding method for v2 protocol and let the configurer choose which use.

For the encoding implementation I've done one for subethamail if you want take a peek. It has been done for testing purposes (so it permits some strange behaviour to check various edge cases and "wrong" headers and can surely be simplified): https://github.com/davidmoten/subethasmtp/blob/master/src/test/java/org/subethamail/smtp/internal/proxy/ProxyProtocolV2HandlerTest.java

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java [hunk 4]
```diff
@@ -143,10 +150,56 @@ protected void initChannel(SocketChannel ch) throws Exception {
...
+                        }
+                    }
+                }
+            }
         });
     }
 
+    private ByteBuf encodeProxyProtocolMessage(HAProxyMessage msg) {
+        // Max length of v1 version proxy protocol message is 108
+        ByteBuf out = Unpooled.buffer(108);
```
[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java:178]: I'm using the netty-haproxy-codec directly. I think netty codec already handled this?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarChannelInitializer.java [hunk 5]
```diff
@@ -118,6 +119,9 @@ protected void initChannel(SocketChannel ch) throws Exception {
             ch.pipeline().addLast("ByteBufPairEncoder", ByteBufPair.ENCODER);
         }
 
+        if (pulsar.getConfiguration().isHaProxyProtocolEnabled()) {
+            ch.pipeline().addLast(OptionalProxyProtocolDecoder.NAME, new OptionalProxyProtocolDecoder());
```
[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarChannelInitializer.java:123]: This is a flag to control if users want to enable the haproxy protocol support. We must provide a way to disable this feature.

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java (hunk 3 above)
Comment (codelipenghui) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java:165]: It's not a required for this PR, we can support the proxy set the different protocol version in the following PRs.

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java:178]: I just noticed the max length for v1 is 108 https://github.com/netty/netty/blob/4.1/codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessageDecoder.java#L37

**Review by sijie — APPROVED** [review]

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java (hunk 3 above)
Comment (diegosalvi) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java:165]: Sounds good

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarChannelInitializer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (diegosalvi) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarChannelInitializer.java:123]: Sure! What I was saying is that there isn't a way to REQUIRE proxy header if pulsar.getConfiguration().isHaProxyProtocolEnabled(). Specs states that is MUST expect a proxy header, it isn't optional. There could be another option to handle optional/mandatory

**Pull request merged**